### DISCUSSION TITLE
One FILTER to one OUTPUT and another FILTER to another output

### DISCUSSION THREAD

**Discussion by vijaynauti** [fluent/fluent-bit#6630] [General]
Hi All

I have configured Fluent-bit for my application deployed on ECS Fargate where i am sending my logs to CloudWatch log and OpenSearch, I have some ECS Metadata attached with my logs like container name and container_ID, these details i need to pass only with OpenSearch but not with CloudWatch log.

Is there is any way to customize this.

Below is my custom configuration.

[INPUT]
    Name forward
    unix_path /var/run/fluent.sock
    Tag test_tag_cw
    Mem_Buf_Limit 100MB
[INPUT]
    Name forward
    unix_path /var/run/fluent.sock
    Tag test_tag_es
    Mem_Buf_Limit 100MB
[FILTER]
    Name record_modifier
    Match test_tag_cw
    Record app-version applog
    Remove_key container_id
[FILTER]
    Name record_modifier
    Match test_tag_es
    Record app-version applog
    Remove_key container_name
[OUTPUT]
    Name cloudwatch
    Match test_tag_cw
    region us-east-1
    log_group_name fluent-bit-cloudwatch-v1
    log_stream_prefix From-fluent-bit-
    auto_create_group true
[OUTPUT]
    Name OpenSearch
    Match test_tag_es
    Host ${my_DNS}
    Port 443
    Index search-logging-v1
    Type _doc
    AWS_Auth On
    AWS_Region us-east-1
    tls On
    Suppress_Type_Name On
    retry_limit 3

**Comment by vijaynauti** [comment]
@PettitWesley

**Reply by PettitWesley** [reply]
> I have some ECS Metadata attached with my logs like container name and container_ID, these details i need to pass only with OpenSearch but not with CloudWatch log.

Does the `log_key` option shown here do what you want? It will only send the value of a single key to CW: https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/cloudwatchlogs

Or if you need something more complicated you might need: https://aws.amazon.com/blogs/opensource/splitting-application-logs-multiple-streams-fluent/

**Reply by vijaynauti** [reply]
@PettitWesley : As mentioned the configuration above, My TAG key with Forward plugin not working.
Once i removed TAG it started working. If that work then i can remove additional log as key field from CloudWatch.

**Reply by PettitWesley** [reply]
Are you using FireLens?

> [INPUT]
Name forward
unix_path /var/run/fluent.sock

This bit of config will be auto-generated and added by FireLens to capture stdout/stderr logs, you don't need to add it yourself: https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#tutorial-replicate-an-ecs-firelens-task-setup-locally

What was your intention with that bit of config?

I do not understand your use case or what you are trying to achieve. Please share more details. Your task def would be ideal too.

**Reply by vijaynauti** [reply]
@PettitWesley : Yes i am using firelens. Application hosted on AWS ECS Fargate and Backend as Java.  I see in documentation that tail plugin support tag whereas i am not sure what log path i need to mentioned hence i am using forward plugin but this plugin does not seems to support tag and Multiline parameter.  

First Query: Can i somehow use tail plugin, if yes then what would be the path (log location) -  I think ECS Fargate sending STDOUT and STDERR to Destination by default,

Second Query: If i can not use tail plugin, then how to use the tag with forward which i have mentioned above.

Third query: In above case which parser entry is getting used in above scenario, as i see multiple entry by default inside parser.conf.

Below is my task def json.   
{
    "taskDefinition": {
        "status": "ACTIVE",
        "memory": "4096",
        "networkMode": "awsvpc",
        "family": "uai200512-smartshop-dev-smartshop",
        "placementConstraints": [],
        "requiresAttributes": [
            {
                "name": "com.amazonaws.ecs.capability.logging-driver.awslogs"
            },
            {
                "name": "ecs.capability.execution-role-awslogs"
            },
            {
                "name": "com.amazonaws.ecs.capability.ecr-auth"
            },
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.19"
            },
            {
                "name": "ecs.capability.firelens.fluentbit"
            },
            {
                "name": "ecs.capability.firelens.options.config.file"
            },
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.17"
            },
            {
                "name": "com.amazonaws.ecs.capability.logging-driver.awsfirelens"
            },
            {
                "name": "com.amazonaws.ecs.capability.task-iam-role"
            },
            {
                "name": "ecs.capability.execution-role-ecr-pull"
            },
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.18"
            },
            {
                "name": "ecs.capability.task-eni"
            }
        ],
        "cpu": "2048",
        "executionRoleArn": "arn:aws:iam::243553853704:role/ecsTaskExecutionRole",
        "compatibilities": [
            "EC2",
            "FARGATE"
        ],
        "volumes": [],
        "requiresCompatibilities": [
            "FARGATE"
        ],
        "taskRoleArn": "arn:aws:iam::243553853704:role/smartshop-ess",
        "taskDefinitionArn": "arn:aws:ecs:us-east-1:243553853704:task-definition/uai200512-smartshop-dev-smartshop:13",
        "containerDefinitions": [
            {
                "dnsSearchDomains": [],
                "environment": [
                    {
                        "name": "app",
                        "value": "smartshop"
                    },
                    {
                        "name": "uai",
                        "value": "uai200512"
                    },
                    {
                        "name": "env",
                        "value": "dev"
                    },
                    {
                        "name": "CloudWatch",
                        "value": "uai200512/smartshop/dev/blobstore-service-lab"
                    }
                ],
                "name": "smartshop",
                "links": [],
                "mountPoints": [],
                "image": "nginx:latest",
                "logConfiguration": {
                    "logDriver": "awsfirelens"
                },
                "dockerLabels": {},
                "cpu": 0,
                "portMappings": [
                    {
                        "protocol": "tcp",
                        "containerPort": 80,
                        "hostPort": 80
                    }
                ],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [],
                "systemControls": [],
                "command": [],
                "essential": true,
                "volumesFrom": []
            },
            {
                "dnsSearchDomains": [],
                "environment": [],
                "firelensConfiguration": {
                    "type": "fluentbit",
                    "options": {
                        "config-file-type": "file",
                        "config-file-value": "/fluent-bit/configs/extra.conf"
                    }
                },
                "systemControls": [],
                "name": "Log_router",
                "links": [],
                "mountPoints": [],
                "image": "243553853704.dkr.ecr.us-east-1.amazonaws.com/logging:fluent-bit",
                "logConfiguration": {
                    "logDriver": "awslogs",
                    "options": {
                        "awslogs-region": "us-east-1",
                        "awslogs-stream-prefix": "From-fluent-bit-",
                        "awslogs-group": "fluent-bit-cloudwatch-v1"
                    }
                },
                "essential": true,
                "portMappings": [],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [
                    "/entrypoint.sh"
                ],
                "user": "0",
                "memory": 180,
                "command": [],
                "cpu": 0,
                "volumesFrom": []
            }
        ],
        "revision": 13
    },
    "tags": []
}

Thanks in Advance..

**Reply by PettitWesley** [reply]
> First Query: Can i somehow use tail plugin, if yes then what would be the path (log location) - I think ECS Fargate sending STDOUT and STDERR to Destination by default,

FireLens does not send STDOUT and STDERR to a file that you can tail. Why do you want to do this?

This is how FireLens sets the tags for logs: https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#firelens-tag-and-match-pattern-and-generated-config

Set your output match patterns based on that. 

Beyond the above, I apologize, its not clear to me what your desired outcome is, I think I need more explanation.

**Reply by vijaynauti** [reply]
Hello @PettitWesley  
I am trying to send logs to OpenSearch and CloudWatch logs by using below custom configuration. My concern is while sending logs to Cloudwatch i don't want and AWS Metadata like ContainerID, Container Name and log as key... If i remove with the help of record_modifier it is removing from OpenSearch as well.  As INPUT section i am using as tail i am not able to tag this. by adding tag key my task not coming up.. I want my Record_modifier to be different for cloudwatch as well as OpenSearch, Hope you got my query. 

[SERVICE]
    flush 2
[INPUT]
    Name forward
    unix_path /var/run/fluent.sock
    Mem_Buf_Limit 100MB
[FILTER]
    Name record_modifier
    Match *
    Remove_key source
[OUTPUT]
    Name cloudwatch
    Match **
    region us-east-1
    log_group_name ${LogGroup_Name}
    log_stream_prefix ${env}-
    auto_create_group true
[OUTPUT]
    Name OpenSearch
    Match  **
    Host ${ES_Host}
    Port 443
    Index smartshop-${Service_Name}-dev-%Y-%m-%d
    Type _doc
    AWS_Auth On
    AWS_Region us-east-1
    tls On
    Suppress_Type_Name On
    retry_limit 3
    Buffer_Size 600KB

**Reply by PettitWesley** [reply]
Does the log_key option for cloudwatch_logs not achieve what you want? https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/cloudwatchlogs#what-if-i-just-want-the-raw-log-line-from-the-container-to-appear-in-cloudwatch

> [INPUT]
Name forward
unix_path /var/run/fluent.sock
Mem_Buf_Limit 100MB

FireLens will auto-add this input. Please do not add it yourself. This will lead to issues if not removed. 

- https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#tutorial-replicate-an-ecs-firelens-task-setup-locally
- https://aws.amazon.com/blogs/containers/how-to-set-fluentd-and-fluent-bit-input-parameters-in-firelens/

**Reply by vijaynauti** [reply]
@PettitWesley  then Mem_Buf_Limit  option if i want to customize, Where should i control this in above case

**Reply by PettitWesley** [reply]
For FireLens, you need to follow:
1. https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/oomkill-prevention#2-mem_buf_limit
2. https://aws.amazon.com/blogs/containers/how-to-set-fluentd-and-fluent-bit-input-parameters-in-firelens/

**Reply by vijaynauti** [reply]
@PettitWesley 

Below is the custom configuration i am using right now..... But issue is while sending logs to CloudWatch i don't want Container_id and container_name, Where while sending logs to OpenSearch i want these parameter to persist there. How should i resolve this.
 
[FILTER]
    Name record_modifier
    Match ss-firelens*
    Remove_key container_id
    Remove_key container_name
    Remove_key ecs_cluster
    Remove_key ecs_task_arn
    Remove_key ecs_task_definition
    Remove_key source
[OUTPUT]
    Name cloudwatch
    Match ss-firelens*
    region us-east-1
    log_group_name fluent-bit-cloudwatch-v1
    log_stream_prefix From-fluent-bit-
    auto_create_group true
[OUTPUT]
    Name OpenSearch
    Match **
    Host *******************
    Port 443
    Index sstest
    Type _doc
    AWS_Auth On
    AWS_Region us-east-1
    tls On
    Suppress_Type_Name On

**Reply by PettitWesley** [reply]
Can you please try this: https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/cloudwatchlogs#what-if-i-just-want-the-raw-log-line-from-the-container-to-appear-in-cloudwatch

And then remove your custom `record_modifier` filter. Opensearch will get all metadata. Cloudwatch will just get the value of the log_key.
